eks: Delete the services associated with the cluster

ricardobranco777 opened the pull request:
This PR:
- Deletes the services associated with a cluster in EKS
- Minor code tidying to avoid indentation & f-strings.

Related ticket: https://progress.opensuse.org/issues/127268

asmorodskyi commented on ocw/lib/eks.py:
```diff
@@ -86,15 +86,15 @@ def kubectl_client(self, region: str, cluster_name: str):
         clusters = {}
         for region in EKS.__cluster_regions:
-            self.log_dbg("Checking clusters in {}", region)
+            self.log_dbg(f"Checking clusters in {region}")
```
I would like to use same style everywhere , if you want to switch to f-strings please prepare commit which do it everywhere not just in few files

ricardobranco777 replied:
> I would like to use same style everywhere , if you want to switch to f-strings please prepare commit which do it everywhere not just in few files

Will do it in this PR then.

ricardobranco777 replied:
> I would like to use same style everywhere , if you want to switch to f-strings please prepare commit which do it everywhere not just in few files

There are far too many lines to change for me to make that change right now.  In some instances the line will exceed the line limit of 130 chars.  So either we do it one module at a time or I rollback the f-string changes in this file.

ricardobranco777 replied:
> I would like to use same style everywhere , if you want to switch to f-strings please prepare commit which do it everywhere not just in few files

These Kubernetes scripts already use f-strings:
- [https://github.com/SUSE/pcw/blob/master/ocw/lib/aks.py](ocw/lib/aks.py)
- [https://github.com/SUSE/pcw/blob/master/ocw/lib/gke.py](ocw/lib/gke.py)
- [https://github.com/SUSE/pcw/blob/master/ocw/lib/k8s.py](ocw/lib/k8s.py)

GCE & Openstack also use them.  I promise to change EC2 & Azure when I touch them.

asmorodskyi replied:
ok

grisu48 commented on ocw/lib/eks.py:
```diff
@@ -105,40 +105,49 @@ def wait_for_empty_nodegroup_list(self, region: str, cluster_name: str, timeout_
+                self.log_dbg(f"Found {len(response['clusters'])} cluster(s) in {region}")
+            for cluster in response['clusters']:
+                resp_nodegroup = self.eks_client(region).list_nodegroups(clusterName=cluster)
+                if len(resp_nodegroup['nodegroups']):
```
```suggestion
                if 'nodegroups' in resp_nodegroup:
```
I think it's better to check for the key in the dict, otherwise we get a `KeyError` if the item is not present.

grisu48 commented on ocw/lib/eks.py:
```diff
@@ -105,40 +105,49 @@ def wait_for_empty_nodegroup_list(self, region: str, cluster_name: str, timeout_
+                        self.log_info(f"Deleting {nodegroup}")
+                        self.eks_client(region).delete_nodegroup(
+                            clusterName=cluster, nodegroupName=nodegroup)
+                if len(resp_nodegroup['nodegroups']):
```
```suggestion
                if 'nodegroups' in resp_nodegroup:
```

asmorodskyi commented on pylintrc:
```diff
@@ -1,5 +1,6 @@
 [MASTER]
 disable=
+ C0103, # Module name "EC2" doesn't conform to snake_case naming style
```
this applies to all additions below

ricardobranco777 replied:
I too disagree, but then we should change the approach to pylint and don't ignore most stuff that is already there.

ricardobranco777 replied:
Fixed.

asmorodskyi commented on Makefile:
```diff
@@ -8,6 +8,10 @@ all: prepare test
 prepare:
 	pip install -r requirements_test.txt
 
+.PHONY: pylint
```
maybe add this to all ?

ricardobranco777 replied:
Fixed.

ricardobranco777 commented:
> I muted some errors in linter because at the point when I was adding linter to project to fix ALL complains we would need to change too big places in code . So I just "make it green" but from current state there is NO options to add more exceptions . For sure not in batch mode . Maybe we can add one or two to this ignore list with really really good explanation why we don't want to obey this rule. I am actually expecting work on REMOVAL lines from this file not adding more things to ignore list

It went from 9.95 to 9.94 because of `ocw/lib/eks.py:119:4: R0912: Too many branches (13/12) (too-many-branches)`

The only way to avoid it would be:

```
-                if 'nodegroups' in resp_nodegroup:
-                    self.log_dbg(f"Found {len(resp_nodegroup['nodegroups'])} nodegroups for {cluster}")
+                self.log_dbg(f"Found {len(resp_nodegroup['nodegroups'])} nodegroups for {cluster}")
```

asmorodskyi commented:
> The only way to avoid it would be:
> 
> ```
> -                if 'nodegroups' in resp_nodegroup:
> -                    self.log_dbg(f"Found {len(resp_nodegroup['nodegroups'])} nodegroups for {cluster}")
> +                self.log_dbg(f"Found {len(resp_nodegroup['nodegroups'])} nodegroups for {cluster}")
> ```


https://pycodequ.al/docs/pylint-messages/r0912-too-many-branches.html , in how to fix it there are lines : 
```
Usually, functions with too many branches fall in one of two categories: (A) functions that do more than one thing or (B) functions that are just a map lookup.
```

I really think that our case is (A) . And actually even without a linter I saw the need to split this big function into smaller ones . e.g. deleting nodes should be separate function . 

P.S. This exactly why I find linter useful and like to follow it complains . Each check didn't come from the void , each check is result of some previous cases where code with certain patterns produced certain problems . So when you get complain from linter it is usually good advice how to improve your code

ricardobranco777 commented:
> > The only way to avoid it would be:
> > ```
> > -                if 'nodegroups' in resp_nodegroup:
> > -                    self.log_dbg(f"Found {len(resp_nodegroup['nodegroups'])} nodegroups for {cluster}")
> > +                self.log_dbg(f"Found {len(resp_nodegroup['nodegroups'])} nodegroups for {cluster}")
> > ```
> 
> https://pycodequ.al/docs/pylint-messages/r0912-too-many-branches.html , in how to fix it there are lines :
> 
> ```
> Usually, functions with too many branches fall in one of two categories: (A) functions that do more than one thing or (B) functions that are just a map lookup.
> ```
> 
> I really think that our case is (A) . And actually even without a linter I saw the need to split this big function into smaller ones . e.g. deleting nodes should be separate function .
> 
> P.S. This exactly why I find linter useful and like to follow it complains . Each check didn't come from the void , each check is result of some previous cases where code with certain patterns produced certain problems . So when you get complain from linter it is usually good advice how to improve your code

Fixed.  I refactored `delete_all_clusters()` and created `delete_nodegrups()` & `delete_services()` along with their respective test functions.

asmorodskyi commented on ocw/lib/eks.py:
```diff
@@ -105,40 +105,54 @@ def wait_for_empty_nodegroup_list(self, region: str, cluster_name: str, timeout_
             self.log_info("Skip waiting due to dry-run mode")
             return None
-        self.log_dbg("Waiting empty nodegroup list in {}", cluster_name)
+        self.log_dbg(f"Waiting empty nodegroup list in {cluster_name}")
```
interesting , this means that when test is running log level is not DEBUG ? :thinking:  Maybe we can tweak log level and increase coverage ? :D

asmorodskyi reviewed (approved):
Perfecto !
